staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,555 @@
+	"github.com/lightningnetwork/lnd/lnrpc/invoicesrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/walletrpc"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+const (
+	defaultConfTarget = 3
+)
+
+// InitHtlcAction is executed if all loop-in information has been validated. We
+// assemble a loop-in request and send it to the server.
+func (f *FSM) InitHtlcAction(_ fsm.EventContext) fsm.EventType {
+	// Lock the deposits and transition them to the LoopingIn state.
+	err := f.cfg.DepositManager.TransitionDeposits(
+		f.loopIn.Deposits, deposit.OnLoopinInitiated, deposit.LoopingIn,
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	// Calculate the swap invoice amount. The pre-pay is added which
+	// effectively forces the server to pay us back our prepayment on a
+	// successful swap.
+	swapInvoiceAmt := f.loopIn.TotalDepositAmount() - f.loopIn.QuotedSwapFee
+
+	// Generate random preimage.
+	var swapPreimage lntypes.Preimage
+	if _, err := rand.Read(swapPreimage[:]); err != nil {
+		log.Error("Cannot generate preimage")
+	}
+	f.loopIn.SwapPreimage = swapPreimage
+	f.loopIn.SwapHash = sha256.Sum256(swapPreimage[:])
+
+	// Derive a client key for the HTLC.
+	keyDesc, err := f.cfg.WalletKit.DeriveNextKey(
+		f.ctx, swap.StaticAddressKeyFamily,
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+	f.loopIn.ClientPubkey = keyDesc.PubKey
+	f.loopIn.HtlcKeyLocator = keyDesc.KeyLocator
+
+	var clientKey [33]byte
+	copy(clientKey[:], keyDesc.PubKey.SerializeCompressed())
+
+	// Create the swap invoice in lnd.
+	_, swapInvoice, err := f.cfg.LndClient.AddInvoice(
+		f.ctx, &invoicesrpc.AddInvoiceData{
+			Preimage:   &swapPreimage,
+			Value:      lnwire.NewMSatFromSatoshis(swapInvoiceAmt),
+			Memo:       "static address loop-in",
+			Expiry:     3600 * 24 * 365,
+			RouteHints: f.loopIn.RouteHints,
+		},
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+	f.loopIn.SwapInvoice = swapInvoice
+
+	f.loopIn.ProtocolVersion = version.AddressProtocolVersion(
+		version.CurrentRPCProtocolVersion(),
+	)
+
+	loopInReq := &loop_rpc.ServerStaticAddressLoopInRequest{
+		SwapHash:         f.loopIn.SwapHash[:],
+		DepositOutpoints: f.loopIn.DepositOutpoints,
+		HtlcClientKey:    f.loopIn.ClientPubkey.SerializeCompressed(),
+		SwapInvoice:      f.loopIn.SwapInvoice,
+		ProtocolVersion:  version.CurrentRPCProtocolVersion(),
+		UserAgent:        f.loopIn.Initiator,
+	}
+	if f.loopIn.LastHop != nil {
+		loopInReq.LastHop = f.loopIn.LastHop
+	}
+
+	loopInResp, err := f.cfg.StaticAddressServerClient.ServerStaticAddressLoopIn( //nolint:lll
+		f.ctx, loopInReq,
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	serverPubkey, err := btcec.ParsePubKey(loopInResp.HtlcServerKey)
+	if err != nil {
+		return f.HandleError(err)
+	}
+	f.loopIn.ServerPubkey = serverPubkey
+
+	// TODO: add cltv validity check.
```

Comment:
I propose to fix this TODO.

Reply:
Repurposed `ValidateLoopInContract` to address this.